Readme.md
# mocha-webdriver
[![NPM version](https://badge.fury.io/js/mocha-webdriver.png)](http://badge.fury.io/js/grunt-mocha-webdriver) [![Build Status](https://travis-ci.org/jmreidy/grunt-mocha-webdriver.svg?branch=master)](https://travis-ci.org/jmreidy/grunt-mocha-webdriver) [![david-dm-status-badge](https://david-dm.org/jmreidy/grunt-mocha-webdriver.png)](https://david-dm.org/jmreidy/grunt-mocha-webdriver#info=dependencies&view=table)
 [![david-dm-status-badge](https://david-dm.org/jmreidy/mocha-webdriver/dev-status.png)](https://david-dm.org/jmreidy/grunt-mocha-webdriver#info=devDependencies&view=table)
> A test harness for running Mocha-based functional tests against
a Webdriver-enabled source: specifically, PhantomJS for local testing and Sauce Labs
for comprehensive cross-browser testing.

This code was extracted from jmreidy/grunt-mocha-webdriver, so that it could be
used independently of Grunt.

##Getting Started
Connecting to Sauce Labs requires java.

Install the plugin with:

```shell
npm install mocha-webdriver --save-dev
```

###Using mocha-webdriver with Phantomjs
Phantom is included via its NPM module. In order
to run tests against phantom, simply add the `usePhantom` flag to the options hash.
The plugin defaults to hitting port 4444, but you can specify your own port via
the `phantomPort` option.

###Using mocha-webdriver with your own Selenium server
 You can run your tests against your own Selenium server instance.
 To do so, use ``hostname`` and ``port`` options.
 Don't forget to remove ``username`` and ``key``.
 Note that the Selenium server should be started and ready before starting the tests.

##Documentation
wd [provides the ability](https://github.com/admc/wd#adding-custom-methods)
to add test methods to its default set of capabilities. `grunt-mocha-webdriver`
exposes the `wd` instance in the same way that `browser` is exposed, so that
you can easily add your own test methods to wd.

Also, Mocha options are exposed to tests. This is especially helpful if you want to reuse the
defined Mocha timeout for your Webdriver tests. For example you can do this in your
Webdriver based E2E tests:

```js
this.browser.waitForElementByCss('.aClass', this.mochaOptions.timeout, cb);
```

Please look at this project's Gruntfile and tests to see all that in action.

###Options
The usual Mocha options are passed through this task to a new Mocha instance.
Please note that while it's possible to specify the Mocha reporter for
tests running on Phantom, there's only one reporter currently supported
for tests against Sauce Labs. This restriction is in place to handle
concurrent Sauce Labs testing sessions, which could pollute the log.

The following options can be supplied to the task:

####usePhantom
Type: Boolean

Specifies whether the task should test against a PhantomJS instance instead
of Sauce Labs. Defaults to false. If true, the tests will run against Phantom
INSTEAD of running against Sauce Labs.

####phantomPort
Type: Int (Default: 4444)

if testing against PhantomJS with the `usePhantom` flag, specify the port
to test against.

####phantomCapabilities
Type: Object (Default: {})

if testing against PhantomJS with the `usePhantom` flag, specify the
[browser capabilities](https://github.com/detro/ghostdriver#what-extra-webdriver-capabilities-ghostdriver-offers).

####phantomFlags
Type: array (Default: [])

if testing against PhantomJS with the `usePhantom` command-line options, specify start additional flags to use. Check [here](http://phantomjs.org/api/command-line.html) or type `phantomjs -h` for complete list of flags.

####usePromises
Type: Boolean

Specifies whether to use the Promise chain version of the WD.js API. Defaults to
false (the callback version).

####ignoreSslErrors
Type: Boolean

A passthrough to the Phantom CLI runner to ignore errors with SSL certs.

####require
Type: Array <String>

An array of paths for requiring before running Mocha tests. Useful for
pre-requires that manipulate Mocha's global environment (e.g.g making Sinon
globally available).

####username
Type: String

The Sauce Labs username to use. Defaults to value of env var `SAUCE_USERNAME`.

####key
Type: String

The Sauce Labs API key to use. Defaults to value of env var `SAUCE_ACCESS_KEY`.

####secureCommands
Type: Boolean (Default: false)

If true, it will use saucelabs, with default `hostname` set to `127.0.0.1` and `port` set to `4445`
in order to send selenium commands through Sauce Connect tunnel (more info
[here](https://saucelabs.com/docs/connect#selenium-relay)).

####autoInstall
Type: Boolean

If `true` this will download Selenium and Chrome Driver to run tests locally.

####hostname
Type: String

If specified, it will connect that selenium server instead of `ondemand.saucelabs.com`.

####port
Type: Int

Selenium server port. Should be used in conjonction with ``hostname``.

####identifier
Type: Number

A Unique identifier for the generated tunnel to Sauce Labs. Will be automatically
generated if not specified. Useful for connected to existing Sauce tunnels.

####concurrency
Type: Int

The number of concurrent browser sessions to spin up on Sauce Labs. Defaults to 1.

####tunnelFlags
Type: Array
An array of option flags for Sauce Connect. See the list of available options
 [here](https://saucelabs.com/docs/connect#connect-flags).

####testName
Type: String

The name of the test, as reported to Sauce Labs.

####testTags
Type: [String]

An array of tags to associate with the test, as reported to Sauce Labs.

####browsers
Type: [Object]

An array of objects specifying which browser options should be passed to Sauce Labs.
Unless a test is run against Phantom (e.g. `usePhantom` is true), this option
*must* be specified. Each browser hash should specify: `browserName`. For saucelabs tests, `platform`,
and `version` are also required.

##Contributing
In lieu of a formal styleguide, take care to maintain the existing coding style. Add
unit tests for any new or changed functionality. Lint and test your code.

##Release History
See History.md

##Contributors
 - Author: [Justin Reidy](https://github.com/jmreidy)
 - You?

##License
Copyright (c) 2014 Justin Reidy

Licensed under the MIT license.


History.md
### v1.0.0
-  Initial port from grunt-mocha-webdriver


index.js
'use strict';

var wd = require('wd');
var SauceTunnel = require('sauce-tunnel');
var _ = require('lodash');
var async = require('async');
var runner = require('./lib/mocha-runner');
var phantom = require('phantomjs');
var childProcess = require('child_process');
var BaseReporter = require('mocha').reporters.Base;
var seleniumLauncher = require('selenium-launcher');
var color = BaseReporter.color;

/*
 * mocha-sauce
 * https://github.com/jmreidy/mocha-sauce
 *
 * Copyright (c) 2014 Justin Reidy
 * Licensed under the MIT license
 */
module.exports = function (files, options) {
    var opts = _.defaults(options, {
      username: process.env.SAUCE_USERNAME,
      key: process.env.SAUCE_ACCESS_KEY,
      identifier: Math.floor((new Date()).getTime() / 1000 - 1230768000).toString(),
      concurrency: 1,
      testName: "",
      testTags: [],
      build: process.env.TRAVIS_BUILD_NUMBER || process.env.BUILD_NUMBER || process.env.BUILD_TAG || process.env.CIRCLE_BUILD_NUM,
      tunnelFlags: null,
      secureCommands: false,
      phantomCapabilities: {},
      phantomFlags: []
    });

    _.forEach(files, function (fileGroup, next) {
      if (opts.usePhantom) {
        runTestsOnPhantom(fileGroup, opts, next);
      }
      else if (opts.autoInstall || opts.hostname && !opts.secureCommands) {
        runTestsOnSelenium(fileGroup, opts, next);
      }
      else {
        runTestsOnSaucelabs(fileGroup, opts, next);
      }
    }, cb);
  });


  function runTestsForBrowser(opts, fileGroup, browser, next) {
    var onTestFinish = function(err) {
      // report mocha test failure
      var callback = function() {
        next(err);
      };
      if (opts.usePromises) {
        browser.quit().nodeify(callback);
      }
      else {
        browser.quit(callback);
      }
    };
    opts.wd = wd;
    runner(opts, fileGroup, browser, grunt, onTestFinish);
  }

  function configureLogEvents(tunnel) {
    var methods = ['write', 'writeln', 'error', 'ok', 'debug'];
    methods.forEach(function (method) {
      tunnel.on('log:'+method, function (text) {
        grunt.log[method](text);
      });
      tunnel.on('verbose:'+method, function (text) {
        grunt.verbose[method](text);
      });
    });
  }

  function runTestsOnPhantom(fileGroup, opts, next) {
    var browser;
    var phantomPort = opts.phantomPort || 4444;
    var phantomCapabilities = opts.phantomCapabilities;
    if (opts.usePromises) {
      browser = wd.promiseChainRemote({port: phantomPort});
    }
    else {
      browser = wd.remote({port: phantomPort});
    }
    grunt.log.writeln('Running webdriver tests against PhantomJS.');

    startPhantom(phantomPort, opts, function (err, phantomProc) {
      if (err) { return next(err); }
      browser.init(phantomCapabilities, function () {
        runTestsForBrowser(opts, fileGroup, browser, function (err) {

          function onClose() {
            grunt.log.writeln('Phantom exited.');
            next(err);
          }

          // the process might already be closed due to an internal crash,
          // so check first is already killed to avoid a deadlock.
          if (phantomProc.killed) {
            onClose();
          }
          else {
            phantomProc.on('close', function() {
              onClose();
            });

            phantomProc.kill();
          }
        });
      });
    });

  }

  function startPhantom(port, opts, next) {
    var phantomOpts = opts.phantomFlags || [];
    phantomOpts.push('--webdriver', port);
    if (opts.ignoreSslErrors) {
      phantomOpts.push('--ignore-ssl-errors', 'yes');
    }
    var phantomProc = childProcess.execFile(phantom.path, phantomOpts);
    var stopPhantomProc = function() {
      phantomProc.kill();
    };
    // stop child phantomjs process when interrupting master process
    process.on('SIGINT', stopPhantomProc);

    phantomProc.on('exit', function () {
      process.removeListener('SIGINT', stopPhantomProc);
    });
    phantomProc.stdout.setEncoding('utf8');
    var onPhantomData = function (data) {
      if (data.match(/running/i)) {
        grunt.log.writeln('PhantomJS started.');
        phantomProc.stdout.removeListener('data', onPhantomData);
        next(null, phantomProc);
      }
      else if (data.match(/error/i)) {
        grunt.log.error('Error starting PhantomJS');
        next(new Error(data));
      }
    };
    phantomProc.stdout.on('data', onPhantomData);
  }

  /**
   * Extracts wd connection params from grunt options
   *
   * Utility function that returns named params
   * that can be used by wd.remote or wd.promiseChainRemote
   */
  function extractConnectionInfo(opts) {
    var params = {};
    var defaultServer = opts.secureCommands ?
                        { hostname: '127.0.0.1', port: 4445 } :
                        { hostname: 'ondemand.saucelabs.com', port: 80 };

    params.hostname = opts.hostname || defaultServer.hostname;
    params.port     = opts.port || defaultServer.port;
    if (opts.key) {
      params.accessKey = opts.key;
    }
    ['auth', 'username'].forEach(function(prop) {
      if (opts[prop]) {
        params[prop] = opts[prop];
      }
    });
    return params;
  }

  /**
   * Init a browser
   */
  function initBrowser(browserOpts, opts, mode, fileGroup, cb) {
    var funcName = opts.usePromises ? 'promiseChainRemote': 'remote';
    var browser = wd[funcName](extractConnectionInfo(opts));
    browser.browserTitle = browserOpts.browserTitle;
    browser.mode = mode;

    browser.mode = mode;
    if (opts.testName) {
      browserOpts.name = opts.testName;
    }
    if (opts.testTags) {
      browserOpts.tags = opts.testTags;
    }
    if (opts.build) {
      browserOpts.build = opts.build;
    }
    if (opts.identifier) {
      browserOpts['tunnel-identifier'] = opts.identifier;
    }

    browser.init(browserOpts, function (err) {
      if (err) {
        grunt.log.error('Could not initialize browser - ' + mode);
        grunt.log.error('Make sure Sauce Labs supports the following browser/platform combo' +
                        ' on ' + color('bright yellow', 'saucelabs.com/platforms') + ': ' + browserOpts.browserTitle);
        return cb(false);
      }
      runTestsForBrowser(opts, fileGroup, browser, cb);
    });
  }

  // used by runTestsOnSaucelabs or runTestsOnSeleni
  var browserFailed = false;

  function pushToQueue(testQueue, browserOpts, browserTitle) {
    testQueue.push(browserOpts, function (err) {
      if (err) {
        browserFailed = true;
      }
      grunt.log.verbose.writeln('%s test complete, %s tests remaining', browserTitle, testQueue.length());
    });
  }

  function startBrowserTests(testQueue, mode, browserOpts) {
    var browserTitle = ''+browserOpts.browserName;
    if (browserOpts.version) {
      browserTitle = browserTitle + ' ' + browserOpts.version;
    }
    if (browserOpts.platform) {
      browserTitle = browserTitle + ' on ' + browserOpts.platform;
    }
    browserOpts.browserTitle = browserTitle;
    grunt.log.verbose.writeln('Queueing ' + browserTitle + ' - ' + mode);
    pushToQueue(testQueue, browserOpts, browserTitle);
  }

  function runTestsOnSaucelabs(fileGroup, opts, next) {
    if (opts.browsers) {
      var tunnel = new SauceTunnel(opts.username, opts.key, opts.identifier, true, opts.tunnelFlags);
      configureLogEvents(tunnel);

      grunt.log.writeln("=> Connecting to Sauce Labs ...");

      tunnel.start(function(isCreated) {
        if (!isCreated) {
          return next(new Error('Failed to create Sauce tunnel.'));
        }
        grunt.log.ok("Connected to Sauce Labs.");

        var testQueue = async.queue(function (browserOpts, cb) {
          // browserOpts, opts, usePromises, errorMsg, fileGroup, cb
          initBrowser(browserOpts,
                      opts,
                      "saucelabs",
                      fileGroup,
                      cb);
        }, opts.concurrency);

        opts.browsers.forEach(function(browserOpts) {
          startBrowserTests(testQueue, 'saucelabs', browserOpts);
        });

        testQueue.drain = function () {
          var err;
          if (browserFailed) {
            err = new Error('One or more tests on Sauce Labs failed.');
          }
          tunnel.stop(function () {
            next(err);
          });
        };
      });
    }
    else {
      grunt.log.writeln('No browsers configured for running on Saucelabs.');
    }
  }

  function runTestsOnSelenium(fileGroup, opts, next) {
    var seleniumServers = [];
    if (opts.browsers) {
      grunt.log.writeln("=> Connecting to Selenium ...");
        var testQueue = async.queue(function (browserOpts, cb) {
          function afterSelenium () {
            var browser = initBrowser(browserOpts,
                                      opts,
                                      "selenium",
                                      fileGroup,
                                      cb);
          }
          if (opts.autoInstall) {
            seleniumLauncher({ chrome: browserOpts.browserName === 'chrome' }, function(err, selenium) {
              seleniumServers.push(selenium);
              grunt.log.writeln('Selenium Running');
              if(err){
                selenium.exit();
                grunt.fail.fatal(err);
                return;
              }
              opts.port = selenium.port;
              afterSelenium();
            }, opts.concurrency);
          } else {
            afterSelenium();
          }
        }, opts.autoInstall ? Object.keys(opts.browsers).length : 1);

      opts.browsers.forEach(function (browserOpts) {
        startBrowserTests(testQueue, 'selenium', browserOpts);
      });

      testQueue.drain = function () {
        var err;
        seleniumServers.forEach(function(seleniumServer) {
          seleniumServer.kill();
        });
        if (browserFailed) {
          err = new Error('One or more tests on Selenium failed.');
        }
        next(err);
      };
    }
    else {
      grunt.log.writeln('No browsers configured for running on Saucelabs.');
    }
  }
};

//wd.js monkey patch for clearer errors
var _newError = wd.webdriver.prototype._newError;
wd.webdriver.prototype._newError = function (opts) {
  var err = _newError(opts);
  try {
    err = new Error(err.cause.value.message
      .match(/([\s\S]*) caused/)[1]
      .match(/'([\s\S]*)'\n/)[1]
    );
  }
  catch (e) {}
  return err;
};
